### PR TITLE
Sync token removal in reorganisation use case

### PR DESCRIPTION
**Pull request by KhalilBellakrid**
Unset sync token when falling in reorg situation because explorers are not trustworthy on that feature
**HODL:** I will add a config to deactivate sync tokens

### CODE REVIEW

File: core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h [hunk 1]
```diff
@@ -230,13 +230,21 @@ namespace ledger {
                 updateCurrentBlock(buddy, account->getContext());
 
                 auto self = getSharedFromThis();
-                return _explorer->startSession().template map<Unit>(account->getContext(), [buddy] (void * const& t) -> Unit {
+                const auto deactivateToken =
+                        buddy->configuration->getBoolean(api::Configuration::DEACTIVATE_SYNC_TOKEN).value_or(false);
+                auto getSyncToken = deactivateToken ? Future<void *>::successful(nullptr) : _explorer->startSession();
+                return getSyncToken.template map<Unit>(account->getContext(), [buddy, deactivateToken] (void * const& t) -> Unit {
```
Comment (alekece) [inline core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h]:
```suggestion
                return getSyncToken.template map<Unit>(account->getContext(), [buddy, deactivateToken] (void * const t) -> Unit {
```

File: core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h [hunk 2]
```diff
@@ -322,52 +330,82 @@ namespace ledger {
...
+                        const auto deactivateToken =
+                                buddy->configuration->getBoolean(api::Configuration::DEACTIVATE_SYNC_TOKEN).value_or(false);
+                        auto startSession = Future<void *>::async(ImmediateExecutionContext::INSTANCE, [=](){
+                            if (deactivateToken) {
+                                return Future<void *>::successful(nullptr);
+                            }
+                            return self->_explorer->startSession();
+                        });
+
+                        return startSession.template flatMap<Unit>(ImmediateExecutionContext::INSTANCE, [=] (void * const &session) {
```
Comment (alekece) [inline core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h]:
```suggestion
                        return startSession.template flatMap<Unit>(ImmediateExecutionContext::INSTANCE, [=] (void * const session) {
```

File: core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h [hunk 3]
```diff
@@ -322,52 +330,82 @@ namespace ledger {
...
 
-                            //Update savedState's batches
-                            for (auto &batch : buddy->savedState.getValue().batches) {
-                                if (batch.blockHeight > lastBlockHeight) {
-                                    batch.blockHeight = (uint32_t)lastBlockHeight;
-                                    batch.blockHash = lastBlockHash;
+                            //Get its block/block height
+                            auto &failedBatch = buddy->savedState.getValue().batches[currentBatchIndex];
+                            auto failedBlockHeight = failedBatch.blockHeight;
+                            auto failedBlockHash = failedBatch.blockHash;
```
Comment (alekece) [inline core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h]:
```suggestion
                            auto const failedBlockHeight = failedBatch.blockHeight;
                            auto const failedBlockHash = failedBatch.blockHash;
```

File: core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h [hunk 4]
```diff
@@ -322,52 +330,82 @@ namespace ledger {
...
                                 }
-                            }
 
-                            //Save new savedState
-                            buddy->preferences->editor()->template putObject<BlockchainExplorerAccountSynchronizationSavedState>(
-                                    "state", buddy->savedState.getValue())->commit();
+                                //Update savedState's batches
+                                for (auto &batch : buddy->savedState.getValue().batches) {
+                                    if (batch.blockHeight > lastBlockHeight) {
+                                        batch.blockHeight = (uint32_t) lastBlockHeight;
```
Comment (alekece) [inline core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h]: We are performing a narrow cast which can result to an unexpected value.

I would like to introduce you `gsl` header-only library. `gsl::narrow_cast` performs some check during the cast and throws a `narrowing_error` if the resulting value is different from the uncasted value.

```suggestion
                                        batch.blockHeight = static_cast<uint32_t>(lastBlockHeight);
```

Comment (KhalilBellakrid) [reply core/src/wallet/common/synchronizers/AbstractBlockchainExplorerAccountSynchronizer.h]: Really cool for `gsl` !! 
As discussed f2f, this is not something changed by this PR, Github is clearly messed up. but still the point is valid and should be treated in a dedicated PR for all our code ! I will create the corresponding ticket.

**Pull request merged**